056-zk-kraft-migration.md

```diff
@@ -0,0 +1,247 @@
+
+The state machine is described in the following diagram:
+
+![Migration Finite State Machine](./images/056-zk-kraft-migration-fsm.png)
+
+The following sections explain the states and transitions in details.
+
+#### Phase 1: Provisioning the KRaft controller quorum
+
+Starting from a `NoMigration` (or null) state (as per `Kafka.status.migrationState` field), the user applies the `strimzi.io/kraft-migration: true` on the `Kafka` custom resource to make the operator aware of the migration "intention".
+The FSM moves to a `WaitControllersInMigration` state where the operator expects the user to create the KRaft controllers pool.
+The user creates the `KafkaNodePool` custom resource containing the definition of the KRaft controllers which are going to replace the ZooKeeper nodes.
+These controllers needed to be deployed in "migration mode", so the `KafkaNodePool` custom resource has to be annotate with `strimzi.io/kraft-migration: dual`.
+
+This way, the operator configures the controllers accordingly with the ZooKeeper migration flag and the connection to the ZooKeeper ensemble. 
+
+```shell
+# Enable the migration
+zookeeper.metadata.migration.enable=true
+
+# ZooKeeper client configuration
+zookeeper.connect=<connection-to-zookeeper>
+```
+
+> NOTE: without an annotation to differentiate the operator behavior, the operator itself just creates the controller nodes in the pool and restarts the Kafka brokers to use them. ZooKeeper is not used anymore but no proper migration takes place (from ZooKeeper to KRaft metadata).
+
+The KRaft controllers should also be provisioned with the same cluster ID as the existing Kafka cluster.
+
+The operator has to wait for the KRaft controllers quorum to be formed, leader elected and the migration process moving into the state of waiting for brokers to register.
+
+The FSM moves to a `WaitBrokersInMigration` state where the operator expects the user to enable ZooKeeper migration on the brokers pool in order to actually start the migration (Phase 2).
+
+From this state it's still possible to not move forward with the migration.
+The user can deleted the `KafkaNodePool` custom resource containing the definition of the KRaft controllers and apply the `strimzi.io/kraft-migration: false` annotation (or delete it) on the `Kafka` custom resource.
+The FSM moves back to the `NoMigration` state (or null) actually exiting from the migration process.
+
+#### Phase 2: Enabling the migration on the brokers
+
+In the `WaitBrokersInMigration` state, the operator expects the user to enable ZooKeeper migration on the brokers pool in order to actually start the migration.
+In order to start the actual migration, the brokers needs to be connected to the KRaft controllers and have the ZooKeeper migration enabled.
+The user annotates the `KafkaNodePool` resource running the current ZooKeeper based brokers with `strimzi.io/kraft-migration: dual`.
```

Comment:
What if there are more of them? How does this sync? Can half of the brokers run in ne mode and other half in the other one?

Reply:
Yeah that's a good point which could be simpler to be addressed if everything is driven via the only `Kafka` resource and not each pool. As I said before I am in the process to get back my first idea using just `Kafka` which I didn't share here.